flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java

```diff
@@ -461,6 +572,56 @@ private void emitRecordWithBroadcastPartitionerOrBroadcastEmitRecord(boolean isB
 		}
 	}
 
+	private void randomEmitMixedBroadcastBufferOrEvent(boolean isBroadcastEvent) throws Exception {
```

Comment:
I'm not sure if I understand the idea behind the unit tests in this file. What's the point of asserting the count and the size of buffers? In doesn't check for the correctness, nor it checks for the performance it just checks for some private implementation detail.

For performance testing we have micro benchmarks.

For correctness we should only check that certain scenarios like, writing x, y, z records or events and/or mixed with broadcasting records/events/watermarks, result with the correct output content of the buffers, regardless of the number/size of the buffers on the output. Otherwise if we optimise/refactor something, all of those tests will start yielding false positive errors.

I'm not sure if you understand my concern/agree with it. If yes, and If you do not want to fix the old tests, could you please at least re-write the new test? This `randomEmitMixedBroadcastBufferOrEvent` test shouldn't care for number of `getNumberOfCreatedBuffers`.

Reply:
My motivation of adding this test is for verifying the logic after switching modes between broadcast and random. That is we have to finish the current `BufferBuilder` once switching modes, and this logic can be reflected on the number of buffers in all the partition queues.

I could understand your concern of  best-effort ignoring the previous tests failure after refactoring something, and I agree with this rule. If we change this new test into data correctness like we already have in `RecordWriterTest#emitOrBroadcastRecord`, it might be difficult to verify how many `LatencyMarker` are emitted actually because this marker would not affect data correctness sometimes. I mean comparing `broadcastEmit` maker with `randomEmit` marker, the results might be same in aspect of data correctness, but the behavior is different.  I would have a try to re-write this test by both verifying data correctness and the number of markers in channel if you think this way make sense.

Reply:
Yes, we can verify the number of sent/received `LatencyMarker`s, I'm fine with that, but I would expect it to be done via deserilizing and not just based on the "number of buffers".

Reply:
Agree